File: Blazor.BrowserAPI/BrowserAPIs/SensorAPI/Sensor/GravitySensor/GravitySensor.js
import { SensorAPI } from "../Sensor";

export class GravitySensorAPI extends SensorAPI {
    /**
     * @param {GravitySensor} gravitySensor
     */
    constructor(gravitySensor) {
        super(gravitySensor);
    }

    /**
     * @param {number} frequency
     * @param {"device" | "screen"} referenceFrame
     * @returns {GravitySensorAPI | null}
     */
    static create(frequency, referenceFrame) {
        if (!("GravitySensor" in window))
            return null;

        /** @type {MotionSensorOptions} */
        let options;
        if (frequency > 0)
            options = { frequency, referenceFrame };
        else
            options = { referenceFrame };

        return new GravitySensorAPI(new GravitySensor(options));
    }


    /**
     * @returns {number}
     */
    getX() {
        return /** @type {GravitySensor} */(this.sensor).x ?? 0;
    }

    /**
     * @returns {number}
     */
    getY() {
        return /** @type {GravitySensor} */(this.sensor).y ?? 0;
    }

    /**
     * @returns {number}
     */
    getZ() {
        return /** @type {GravitySensor} */(this.sensor).z ?? 0;
    }
}

File: Blazor.BrowserAPI/BrowserAPIs/SensorAPI/Sensor/GravitySensor/GravitySensor.cs
using AutoInterfaceAttributes;
using Microsoft.JSInterop;
using System.Diagnostics.CodeAnalysis;

namespace BrowserAPI.Implementation;

/// <summary>
/// <para>The <i>GravitySensor</i> interface of the Sensor APIs provides on each reading the gravity applied to the device along all three axes.</para>
/// <para>To use this sensor, the user must grant permission to the 'accelerometer' device sensor through the Permissions API.</para>
/// <para>Objects of this class must disposed manually, so do not forget to call DisposeAsync() when you are done with it.</para>
/// </summary>
[AutoInterface(Namespace = "BrowserAPI", Inheritance = [typeof(ISensor)])]
[RequiresUnreferencedCode("Uses Microsoft.JSInterop functionalities")]
public sealed class GravitySensor(IJSObjectReference gravitySensor) : Sensor, IGravitySensor {
    private protected override IJSObjectReference SensorJS => gravitySensor;


    /// <summary>
    /// Returns a double containing the acceleration of the device along the device's x axis.
    /// </summary>
    public ValueTask<double> X => GetX(default);

    /// <summary>
    /// Returns a double containing the acceleration of the device along the device's x axis.
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public ValueTask<double> GetX(CancellationToken cancellationToken) => gravitySensor.InvokeTrySync<double>("getX", cancellationToken);


    /// <summary>
    /// Returns a double containing the acceleration of the device along the device's y axis.
    /// </summary>
    public ValueTask<double> Y => GetY(default);

    /// <summary>
    /// Returns a double containing the acceleration of the device along the device's y axis.
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public ValueTask<double> GetY(CancellationToken cancellationToken) => gravitySensor.InvokeTrySync<double>("getY", cancellationToken);


    /// <summary>
    /// Returns a double containing the acceleration of the device along the device's z axis.
    /// </summary>
    public ValueTask<double> Z => GetZ(default);

    /// <summary>
    /// Returns a double containing the acceleration of the device along the device's z axis.
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public ValueTask<double> GetZ(CancellationToken cancellationToken) => gravitySensor.InvokeTrySync<double>("getZ", cancellationToken);
}
